### PR TITLE
Ltac2: add notations for eval cbv in ... and other in place reductions

### PR DESCRIPTION
**Pull request by MSoegtropIMC**
This PR adds notations for tactics like "eval cbv in".

One open question: to which tactic level should I assign the notations. What is the default if I don't give any? The test cases and my private use work fine without giving a level, but I would like to understand this.

**Kind:** feature.
- [x] Added / updated test-suite
- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by JasonGross** [comment]
> One open question: to which tactic level should I assign the notations. What is the default if I don't give any? The test cases and my private use work fine without giving a level, but I would like to understand this.

I'm not sure what the default is in Ltac2, but it ought to be level 1, I think.  See https://coq.github.io/doc/master/refman/proof-engine/ltac.html#grammar-token-ltac-expr for the various levels that are used in Ltac1.

File: test-suite/ltac2/notations.v
Comment (JasonGross) [inline test-suite/ltac2/notations.v]: Do we need the outer parentheses here?  If not, we should drop them.

File: test-suite/ltac2/notations.v
Comment (JasonGross) [inline test-suite/ltac2/notations.v]: These should be in the output test-suite, to make sure that the notations are binding the right behavior.

File: test-suite/ltac2/notations.v
Comment (JasonGross) [inline test-suite/ltac2/notations.v]: Where is `print_sc` defined?

File: user-contrib/Ltac2/Notations.v
Comment (JasonGross) [inline user-contrib/Ltac2/Notations.v]: How does this work?  I don't think `simpl` takes `beta iota zeta delta` arguments like all the other tactics.....

File: user-contrib/Ltac2/Notations.v
Comment (JasonGross) [inline user-contrib/Ltac2/Notations.v]:
```suggestion
Ltac2 Notation "eval" "vm_compute" pl(opt(seq(pattern, occurrences))) "in" c(constr) :=
```
I don't think the vm supports patterns nor occurences though....

File: user-contrib/Ltac2/Notations.v
Comment (JasonGross) [inline user-contrib/Ltac2/Notations.v]:
```suggestion
Ltac2 Notation "eval" "native_compute" pl(opt(seq(pattern, occurrences))) "in" c(constr) :=
```
I don't think the native compiler supports patterns nor occurences though....

**Review by JasonGross — CHANGES_REQUESTED** [review]
I don't think this is the right way to go about doing things.  `eval ... in ...` is supposed to support any [`red_expr`](https://coq.github.io/doc/master/refman/language/gallina-specification-language.html#grammar-token-red-expr).  Note that this also includes things like `Declare Reduction ident := ...`.  We should have a parsing scope for `red_expr`, and should just use that.  I acknowledge that this is a deeper change, though, and this PR is better than what we have now, so I will not block it if it's not clear how to support `red_expr` in general. However, I think we should at least add a comment that the current notations are temporary while waiting for support for `red_expr`.

File: user-contrib/Ltac2/Notations.v
Comment (JasonGross) [reply user-contrib/Ltac2/Notations.v]: > I don't think the vm supports patterns nor occurences though....

Oh, huh, I guess it does.  I never knew that.

File: user-contrib/Ltac2/Notations.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Notations.v]: well the Ltac2 version does. But I can't tell what it actually does with it, since I never user simpl (switched to cbn years ago and never looked back). I will try to find out.

**Comment by MSoegtropIMC** [comment]
> I don't think this is the right way to go about doing things. `eval ... in ...` is supposed to support any [`red_expr`](https://coq.github.io/doc/master/refman/language/gallina-specification-language.html#grammar-token-red-expr).

Well isn't this what the "strategy" scope is?

> Note that this also includes things like `Declare Reduction ident := ...`

IMHO this doesn't make that much sense in Ltac2. In Ltac2 a reduction is a simple Ltac2 record (`Ltac2 Type red_flags`) and I wouldn't find it very orthogonal to have a special syntax to define variables of this type. If you have such a record, using e.g. `Std.eval_cbn` right away is sufficiently straight forward. These notations are for the case where you want to give the the reduction strategy in place, which is quite painful as is.

But one could provide a notation for creating a reduction record using the comfortable strategy scope - I will do this.

Btw.: this is all the same with the existing cbv, cbn, ... tactics which work on the goal rather than on a given term.

**Comment by MSoegtropIMC** [comment]
> > One open question: to which tactic level should I assign the notations. What is the default if I don't give any? The test cases and my private use work fine without giving a level, but I would like to understand this.
> 
> I'm not sure what the default is in Ltac2, but it ought to be level 1, I think. See https://coq.github.io/doc/master/refman/proof-engine/ltac.html#grammar-token-ltac-expr for the various levels that are used in Ltac1.

Thank you for the pointer. I think the level numbers are not the same in Ltac2, but it definitely helps to know what level it should be in!

**Comment by JasonGross** [comment]
> > Note that this also includes things like `Declare Reduction ident := ...`
> 
> IMHO this doesn't make that much sense in Ltac2.

Note that reduction expressions are not just about Ltac1, but also, e.g., the vernacular `Eval ... in ...`.

**Comment by MSoegtropIMC** [comment]
> Note that reduction expressions are not just about Ltac1, but also, e.g., the vernacular `Eval ... in ...`.

Well I am still not a big fan if it. The advantage of the Ltac2 way is that you can create a computed reduction strategy, notably with computed symbol lists for delta. You can pass them around in function calls, ... . I think @ppedrot needs to decide if this should be implemented in the strategy context (btw: maybe it even is - did you check?).

But again: I think this is an extended topic. Here I jsut extend what already exists for the goal oriented reductions to the eval style reductions. If both shall be extended, this is a different thing.

**Comment by MSoegtropIMC** [comment]
@JasonGross : how do we conclude here? My suggestion is that we integrate support for `Decalare Reduction` in a separate step, if people really feel that this is required. My opinion is that the mechanism provided by Ltac2 (just define a `red_flags` record as variable or provide a function which computes one) is more flexible and more orthogonal. But of cause it does not hurt to integrate for compatibility reasons support for `Declare Reduction`. I don't think it would be a big deal. But I honestly would wait until there are real requests for it. What I am not sure about: do you feel you need this or was this a more general comment? I am actually using computed `red_flags` records in one place and a few defined ones.

Besides support for `Declare Reduction`, do you see any other issues with this PR - besides the naming for `vm_compute` and `native_compute`?

What I will add is a notation for creating `ref_flags` records using the sugar syntax. Do you have a suggestion for a name? How about `define_reduction ...`. I feel just `reduction` is a bit too name space invasive.

**Comment by ppedrot** [comment]
I'd advise against cluttering the default Ltac2 syntax with many notations when their usefulness is dubious. The only reason `eval` existed in Ltac1 was that value-returning functions had to be hardwired in the syntax. So, I believe it's fine to provide the user backward-compatible syntaxes, but they should be opt-in, e.g. defined inside a specific module.

**Comment by MSoegtropIMC** [comment]
> The only reason `eval` existed in Ltac1 was that value-returning functions had to be hardwired in the syntax.

OK then I am missing something. How would I reduce / simplify a term in Ltac2 without using the eval functions? And since writing red_flags records literally is not much of an option how would one do this without these notations? I would say about 10%..20% of the tactics I call are eval tactics of this kind. I needs this really frequently, e.g. before one matches a constr term in Ltac2, one has to make sure that it has the properly reduced form. Unlike a match in gallina Ltac2 does not do reductions when matching a constr with a pattern in a match!. But even if there would be a more gallina like match_unify! or whatever, I still have plenty of places where I do an eval.

**Comment by MSoegtropIMC** [comment]
@ppedrot : can you please explain why you think "eval ... in" is not required in Ltac2 and just nice to have for syntactic compatibility with Ltac1? See also my previous post. If this helps I can post a few typical usage examples. As I said I cannot imagine how to live without this - short of reimplementing the various eval methods in Ltac2.

**Comment by ppedrot** [comment]
@MSoegtropIMC So, you're right that it's handy, but I was under the impression that the most standard uses could be replaced by mere function calls. I was probably wrong, although I am still cautious about overuses of notations, so since we're not bound to 100% backwards compatibility, we could go for a syntax that guarantees modularity, and provide the old syntax via a separate import.

**Comment by MSoegtropIMC** [comment]
@ppedrot : can you give a few hints on what syntax you have in mind? One could have a notation for creation of `red_flags` structures, which I would anyway suggest, and then use this for cbv, cbn and lazy.  But providing the input for e.g. unfold is also not terribly easy because depending on where the symbol comes from, one needs to construct the references in a different way. I used to setup red_flags records with delta list manually for a while and I think I used 3 or 4 different ways to construct references. And on top come the occurances.

Honestly IMHO the simplest and most canonical way to do this is to have exactly the same syntax as in the goal oriented tactics and in Ltac1 and this is what I do here. The syntax of all these tactics is quite complicated (even @JasonGross wrote above that he didn't know some details of it) and this way it is documented in the references manual. If we do it in a different way, the documentation allone would be quite a bit of work.

**Comment by JasonGross** [comment]
> how do we conclude here? My suggestion is that we integrate support for `Decalare Reduction` in a separate step, if people really feel that this is required.

Indeed, this PR is clearly an improvement over the current state, and thus I support it.  Note, though, that your argument seems to be pulling in the opposite direction of #11679.
I would like to see things go in an even more radical direction, where the language for specifying reductions gains significantly more expressivity (so that I can, e.g., say that arguments to `foo` should be fully reduced with `vm_compute`, arguments to `bar` should be fully reduced with `native_compute`, and `foo` and `bar` themselves should be unfolded with `cbv`, and arguments to `baz` should be reduced with `cbn [fst snd]`, except that things occurring as a subterm to an argument to `baz` should not be inspected for reduction at all).

**Comment by ppedrot** [comment]
The bottom line is that I am OK with this PR, even in its current state. @JasonGross do you want to assign and merge?

**Comment by JasonGross** [comment]
(Sorry, just saw the mention now) I am willing to assign and merge, but I am still waiting for @MSoegtropIMC to implement my requested changes (primarily, `eval vm` -> `eval vm_compute`, `eval native` -> `eval native_compute`, and move the tests to the output test-suite).

**Comment by JasonGross** [comment]
Also, this maybe deserves a changelog entry?

**Comment by MSoegtropIMC** [comment]
@JasonGross : thanks, I will do both today.

**Comment by MSoegtropIMC** [comment]
@JasonGross : sorry I was busy with other things which took longer than expected.

One question: should I create a notation to define `red_flags` records using the simplified syntax for storing in a Ltac2 variable. If so, do you have a suggestion for the syntax?

**Comment by JasonGross** [comment]
> One question: should I create a notation to define `red_flags` records using the simplified syntax for storing in a Ltac2 variable. If so, do you have a suggestion for the syntax?

Do you mean something like `(beta iota delta [Nat.add])%strategy`, or something else?

**Comment by MSoegtropIMC** [comment]
> Do you mean something like (beta iota delta [Nat.add])%strategy (or is it strategy:(beta iota delta [Nat.add])?), or something else?

Yes something like this - I like the `strategy:` variant. But the main question, also @ppedrot is if you think we should have this or not, or if we should wait until the reduction tactics can take a reduction variable. If I define this, once could use this only for the core non notation functions which is half way inconsistent.

**Comment by Zimmi48** [comment]
@coq/ltac2-maintainers what is the status of this? More discussion needed or this can be merged soon? Please adjust the milestone as appropriate.

**Comment by MSoegtropIMC** [comment]
It is almost finished - I was draged away by other things. I will finish it next week end.

**Comment by Zimmi48** [comment]
If the PR is still in progress, then it seems a bit late for 8.12, especially as the branching is expected on Friday.

**Comment by MSoegtropIMC** [comment]
It is more editorial.

**Comment by Zimmi48** [comment]
If @JasonGross and @ppedrot are OK, it's good for us if it's merged over the week-end. Not later than this though.

**Comment by JasonGross** [comment]
I am happy to merge this as soon as my two requests are addressed (renaming the VM and native functions to be more conventional, and adding a changelog).  Anything else is extra.

**Comment by MSoegtropIMC** [comment]
> I am happy to merge this as soon as my two requests are addressed (renaming the VM and native functions to be more conventional, and adding a changelog). Anything else is extra.

Thanks - I have a bit of a backlog. I try to do this until Monday.

**Comment by ejgallego** [comment]
@JasonGross @MSoegtropIMC branching for 8.12 will happen pretty soon, if you can get this merged at the latest on Sunday we are happy to include it in 8.12 [as @ppedrot did approve long time ago] , however if you need more days I'm afraid this will miss 8.12.

**Comment by MSoegtropIMC** [comment]
@ejgallego : I can do it tomorrow - also the PR on Ltac2 bools, which I think is even more important. Not sure if @JasonGross can merge it on Sunday, though.

**Comment by JasonGross** [comment]
Let me know when to want me to re-review and merge

File: test-suite/ltac2/notations.v
Comment (MSoegtropIMC) [reply test-suite/ltac2/notations.v:28]: I moved it there but left a note here (for me the test cases are sometimes the only doc and one wouldn't find it otherwise).

**Comment by MSoegtropIMC** [comment]
@JasonGross : it should be good now. I did the name changes, add a changelog and also moved the  test to the output section, but I left a note in the ltac2 specific file.

CI seems bogus (does not start) - I will restart it.

**Comment by MSoegtropIMC** [comment]
@JasonGross : as far as I can tell the two failing CI tests also fail in master, e.g.:
(https://gitlab.com/coq/coq/pipelines/146717553)

File: doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst
Comment (JasonGross) [inline doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst:2]: I don't think this is how it gets cited.  @coq/doc-maintainers, how is this supposed to be formatted?

**Review by JasonGross — APPROVED** [review]
LGTM modulo the changelog entry.  I'll merge tomorrow (and open a new PR for the changelog entry fixing if need be)

File: doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst
Comment (MSoegtropIMC) [reply doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst:2]: Knowing nothing about the RST format, I copied that from ltac.rst, assuming that all .rst files accept the same input format. The result of ltac.rst can be seen here (https://coq.github.io/doc/master/refman/proof-engine/ltac.html#computing-in-a-constr). But the issue with this is that it would link to wrong syntax definitions. I try to find the right ones.

Comment (MSoegtropIMC) [reply doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst]: @JasonGross : I now looked through the compiled manual for the kind of formatting which is appropriate and then looked at the sources for the RST syntax. As far as I can tell now it is what is standard for inline references to syntax. The format I had before was indeed more a definition display format. I also changed the reference to ltac2_term, but I kept the reference to the ltac reduction definition, since it should be the same.

The examples I chose are the inline syntax definitions in the first text paragraph under (https://coq.github.io/doc/master/refman/proof-engine/ltac.html#proving-a-subgoal-as-a-separate-lemma).

**Review by JasonGross — APPROVED** [review]

**Comment by MSoegtropIMC** [comment]
@JasonGross : FYI: the two failing CI items also fail in master (at least sometimes). See e.g. (https://gitlab.com/coq/coq/pipelines/146717553)

File: doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst
Comment (Zimmi48) [reply doc/changelog/05-tactic-language/11981-ltac2-eval-notations.rst]: I confirm that this is good now.

**Comment by Zimmi48** [comment]
The CI errors are because of some unmerged overlays. One of them has been merged since then, so I restarted that one. The other, you can simply ignore.

**Pull request merged**
